snuba/query/processors/mapping_optimizer.py

```diff
@@ -0,0 +1,192 @@
+        ):
+            rhs = match.expression("right_hand_side")
+            assert isinstance(rhs, LiteralExpr)
+            return (
+                ConditionClass.NOT_OPTIMIZABLE
+                # ifNull(tags[asd], '') = '' is not optimizable.
+                if rhs.value == ""
+                else ConditionClass.OPTIMIZABLE
+            )
+        elif match is None:
+            # If this condition is not matching an optimizable condition,
+            # check that it does not reference the optimizable column.
+            # If it does, it means we should not optimize this query.
+            for exp in condition:
+                if isinstance(exp, Column) and exp.column_name in (
+                    f"{self.__column_name}.key",
+                    f"{self.__column_name}.value",
+                ):
+                    return ConditionClass.NOT_OPTIMIZABLE
+            return ConditionClass.IRRELEVANT
+        else:
+            return ConditionClass.IRRELEVANT
+
+    def __replace_with_hash(self, condition: Expression) -> Expression:
+        match = self.__optimizable_pattern.match(condition)
+        if (
+            match is None
+            or match.string(KEY_COL_MAPPING_PARAM) != f"{self.__column_name}.key"
+        ):
+            return condition
+        rhs = match.expression("right_hand_side")
+        assert isinstance(rhs, LiteralExpr)
+        key = match.string(KEY_MAPPING_PARAM).translate(ESCAPE_TRANSLATION)
+        return FunctionExpr(
+            alias=condition.alias,
+            function_name="has",
+            parameters=(
+                Column(
+                    alias=None,
+                    table_name=match.optional_string(TABLE_MAPPING_PARAM),
+                    column_name=self.__hash_map_name,
+                ),
+                FunctionExpr(
+                    alias=None,
+                    function_name="cityHash64",
+                    parameters=(LiteralExpr(None, f"{key}={rhs.value}"),),
+                ),
+            ),
+        )
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        if not get_config("tags_hash_map_enabled", 0):
+            return
+
+        cond_class = ConditionClass.IRRELEVANT
+        condition = query.get_condition_from_ast()
+        if condition is not None:
+            cond_class = self.__classify_combined_conditions(condition)
+            if cond_class == ConditionClass.NOT_OPTIMIZABLE:
+                return
+
+        having_cond_class = ConditionClass.IRRELEVANT
+        having_cond = query.get_having_from_ast()
+        if having_cond is not None:
+            having_cond_class = self.__classify_combined_conditions(having_cond)
+            if having_cond_class == ConditionClass.NOT_OPTIMIZABLE:
+                return
+
+        if (
+            cond_class != ConditionClass.OPTIMIZABLE
+            and having_cond_class != ConditionClass.OPTIMIZABLE
+        ):
```

Comment:
Personal preference / minor style comment: I find `not (a or b)` easier to read than `(!a and !b)`